### PR TITLE
Remove usage of CommaExp from dmd

### PR DESCRIPTION
**Pull request by mathias-lang-sociomantic**
Split from https://github.com/dlang/dmd/pull/5737

### CODE REVIEW

File: src/dtemplate.d [hunk 1]
```diff
@@ -3784,8 +3784,10 @@ extern (C++) MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplatePara
                     RootObject o2 = (*tp.tempinst.tiargs)[i];
                     Type t2 = isType(o2);
 
-                    size_t j;
-                    if (t2 && t2.ty == Tident && i == tp.tempinst.tiargs.dim - 1 && (j = templateParameterLookup(t2, parameters), j != IDX_NOTFOUND) && j == parameters.dim - 1 && (*parameters)[j].isTemplateTupleParameter())
+                    size_t j = (t2 && t2.ty == Tident && i == tp.tempinst.tiargs.dim - 1)
+                        ? templateParameterLookup(t2, parameters) : IDX_NOTFOUND;
+                    if (j != IDX_NOTFOUND && j == parameters.dim - 1
+                        && (*parameters)[j].isTemplateTupleParameter())
```
Comment (yebblies) [inline src/dtemplate.d]: DMD style puts `&&` on the right.

File: src/expression.d [hunk 2]
```diff
@@ -12795,13 +12795,11 @@ public:
...
 
-                    CallExp ce;
-                    DotVarExp dve;
-                    if (sd.ctor &&
-                        e2y.op == TOKcall &&
-                        (ce = cast(CallExp)e2y, ce.e1.op == TOKdotvar) &&
-                        (dve = cast(DotVarExp)ce.e1, dve.var.isCtorDeclaration()) &&
-                        e2y.type.implicitConvTo(t1))
+                    CallExp ce = (e2y.op == TOKcall) ? cast(CallExp)e2y : null;
+                    DotVarExp dve = (ce !is null && ce.e1.op == TOKdotvar)
```
Comment (yebblies) [inline src/expression.d]: Can just be `ce && ...`

File: src/dtemplate.d (hunk 1 above)
Comment (mathias-lang-sociomantic) [inline src/dtemplate.d]: IMO that's much less readable, but if it's the existing style...
Now also seems a good time to reiterate the need of a `CONTRIBUTING.md` file :)

**Comment by mathias-lang-sociomantic** [comment]
Updated:

```diff
diff --git a/src/dtemplate.d b/src/dtemplate.d
index 45be215..e262dae 100644
--- a/src/dtemplate.d
+++ b/src/dtemplate.d
@@ -3786,8 +3786,8 @@ extern (C++) MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplatePara
 
                     size_t j = (t2 && t2.ty == Tident && i == tp.tempinst.tiargs.dim - 1)
                         ? templateParameterLookup(t2, parameters) : IDX_NOTFOUND;
-                    if (j != IDX_NOTFOUND && j == parameters.dim - 1
-                        && (*parameters)[j].isTemplateTupleParameter())
+                    if (j != IDX_NOTFOUND && j == parameters.dim - 1 &&
+                        (*parameters)[j].isTemplateTupleParameter())
                     {
                         /* Given:
                          *  struct A(B...) {}
diff --git a/src/expression.d b/src/expression.d
index 2e195d6..63a0b9a 100644
--- a/src/expression.d
+++ b/src/expression.d
@@ -12796,10 +12796,10 @@ public:
                         e2y = (cast(CommaExp)e2y).e2;
 
                     CallExp ce = (e2y.op == TOKcall) ? cast(CallExp)e2y : null;
-                    DotVarExp dve = (ce !is null && ce.e1.op == TOKdotvar)
+                    DotVarExp dve = (ce && ce.e1.op == TOKdotvar)
                         ? cast(DotVarExp)ce.e1 : null;
-                    if (sd.ctor && ce && dve && dve.var.isCtorDeclaration()
-                        && e2y.type.implicitConvTo(t1))
+                    if (sd.ctor && ce && dve && dve.var.isCtorDeclaration() &&
+                        e2y.type.implicitConvTo(t1))
                     {
                         /* Look for form of constructor call which is:
                          *    __ctmp.ctor(arguments...)
diff --git a/src/func.d b/src/func.d
index cfa45fe..ba4836f 100644
--- a/src/func.d
+++ b/src/func.d
@@ -1275,9 +1275,9 @@ public:
          * the function body.
          */
         TemplateInstance ti;
-        if (fbody
-            && (isFuncLiteralDeclaration() || (storage_class & STCinference) || (inferRetType && !isCtorDeclaration()) || isInstantiated() && !isVirtualMethod()
-                && ((ti = parent.isTemplateInstance()) is null || ti.isTemplateMixin() || ti.tempdecl.ident == ident)))
+        if (fbody &&
+            (isFuncLiteralDeclaration() || (storage_class & STCinference) || (inferRetType && !isCtorDeclaration()) || isInstantiated() && !isVirtualMethod() &&
+             ((ti = parent.isTemplateInstance()) is null || ti.isTemplateMixin() || ti.tempdecl.ident == ident)))
         {
             if (f.purity == PUREimpure) // purity not specified
                 flags |= FUNCFLAGpurityInprocess;
```

File: src/dtemplate.d (hunk 1 above)
Comment (yebblies) [inline src/dtemplate.d]: There's a PR somewhere about auto-formatting the source, we'll get there one day.

**Comment by mathias-lang-sociomantic** [comment]
Anything else I need to do on my end ?

**Comment by yebblies** [comment]
Auto-merge toggled on

**Pull request merged**